src/circuit_builder.rs

```diff
@@ -26,21 +26,71 @@ pub enum PublicInput {
 }
 
 impl PublicInput {
-    /// Returns the value of a PublicInput struct
-    pub fn value(&self) -> Vec<BlsScalar> {
+    /// Returns the serialized-size of the `PublicInput` structure.
+    pub const fn serialized_size() -> usize {
+        33usize
+    }
+
+    /// Returns the byte-representation of a [`PublicInput`].
+    /// Note that the underlying variants of this enum have different
+    /// sizes on it's byte-representation. Therefore, we need to return
+    /// the biggest one to set it as the default one.
+    pub fn to_bytes(&self) -> [u8; Self::serialized_size()] {
+        let mut bytes = [0u8; Self::serialized_size()];
         match self {
-            PublicInput::BlsScalar(scalar, _) => vec![*scalar],
-            PublicInput::JubJubScalar(scalar, _) => vec![BlsScalar::from(*scalar)],
-            PublicInput::AffinePoint(point, _, _) => vec![point.get_x(), point.get_y()],
+            Self::BlsScalar(scalar, _) => {
+                bytes[0] = 0;
+                bytes[1..33].copy_from_slice(&scalar.to_bytes());
+                bytes
+            }
+            Self::JubJubScalar(scalar, _) => {
+                bytes[0] = 1;
+                bytes[1..33].copy_from_slice(&scalar.to_bytes());
+                bytes
+            }
+            Self::AffinePoint(point, _, _) => {
+                bytes[0] = 2;
+                bytes[1..Self::serialized_size()].copy_from_slice(&point.to_bytes());
+                bytes
+            }
+        }
+    }
+
+    /// Generate a [`PublicInput`] structure from it's byte representation.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, CircuitErrors> {
+        if bytes.len() != Self::serialized_size() {
+            return Err(CircuitErrors::InvalidPublicInputBytes.into());
+        } else {
+            let mut array_bytes = [0u8; 32];
+            array_bytes.copy_from_slice(&bytes[1..Self::serialized_size()]);
+            match bytes[0] {
+                0 => Ok(Self::BlsScalar(
```

Comment:
I think you should be able to collect the result of the `from_bytes()` and then do the `Option::from(result).ok_or_else(|| …)?` just once.

Reply:
Ohh ye. I missed it. Good point

Reply:
I don't think so. 
I need to call the `ok_or_else` inside the enum variant generation. I can't unify this after the match since the `match` will be returning different types.

Maybe I'm missing something. But it doesn't seem to be possible. It would be different if the options were retutning the enum variant directly. Then the match arms will always return the same type.

Reply:
No, you're right I missed the part where you need to returns different enum variant, it's a shame but we cannot easily reduce the boilerplate here.